### ISSUE TITLE
[Proposal] JSON-RPC API

### ISSUE THREAD

**Issue by zone117x** [blockstack/stacks-blockchain#970]
_Note: descriptions, names, types, formatting -- all WIP_

# Methods (prod)
* `call_func`(contract identifier, function name, function args) -> ok(function return data), err(function_not_found, parser, type checker, or runtime error)
* `deploy_contract`(contract identifier, contract source code) -> ok(contract address), err(parser, type checker, or runtime error)
* `get_deployment_cost`(contract source code) -> ok(stacks tokens int), err(parser, or type checker error)
* `send_transaction`(contract identifier, function name, function args) -> ok(tx_receipt ?), err(type checker or runtime error)
* `get_map_entry`(contact identifier, map name, map key value) -> ok(value data), err()
* `get_contract_source`(contract identifier) -> ok(source string), err(contract_not_found)

# Methods (dev mode only)
* `dev_mine_block`(optional unix timestamp) -> ok(block number)

# Object Types

### Responses
* Errors
  * parse_error(msg) 
  * type_error(msg)
  * runtime_error(msg)
  * map_not_found
  * map_entry_not_found
  * contract_not_found
  * function_not_found
* tx_receipt(cost, tx_hash)

**Comment by zone117x** [comment]
Question:
Deploying a contract and transacting with a contract are both "just transactions", but I think its useful to have separate APIs for them. 
What is an explicit & useful naming convention? For example `tx_deploy_contract` and `tx_invoke_contract` ?

**Comment by hstove** [comment]
Why the difference between `call_func` and `send_transaction`? I assume `call_func` is a "read-only" operation, like `call` in ETH. `send_transaction` seems like a bad name if it's meant to execute a contract function, and `send_transaction` should be the RPC method for broadcasting a normal STX transfer.

> What is an explicit & useful naming convention? For example `tx_deploy_contract` and `tx_invoke_contract` ?

Related to my point above, `invoke_contract` is better, and I kind of like these without the `tx` prefix, although I understand it makes it more explicit.

> call_func(contract identifier, function name, function args) -> ok(function return data), err(function_not_found, parser, type checker, or runtime error)

How can you get a parser error here? Shouldn't the contract not exist if there is a parsing error? Or could you get a situation where a non-parsable contract exists?

> dev_mine_block(optional unix timestamp) -> ok(block number)

Probably should also pass an address for mining rewards?

I think it would be great to further flesh out the dev mode API, maybe even for minting funds to an address.

**Comment by hstove** [comment]
> `send_transaction(contract identifier, function name, function args)`

Forgive me, I'm more out of the loop on our smart contracts. Do we have the concept of passing a `value`, similar to ETH? Or would you just pass some `int` argument to do things like send funds to a contract?

**Comment by jcnelson** [comment]
Yeah, I'm a little confused by what is meant by `call_func`.  In particular, I'd like to avoid any API calls that cause the node to execute code for free.  If the purpose of `call_func` is to evaluate a function with respect to the global prod context, then the client should run the VM locally but back reads to the contract DB to the node.  That way, the node only services read requests (which it has to anyway), and the client takes full responsibility for running code and storing intermediate results.

There won't be a difference between deploying a contract and sending a transaction -- it's all the same to the node.  There probably should just be a single `broadcast_transaction`-like endpoint that takes a signed and serialized transaction as input, adds it to its mempool, and relays it to its neighbors.

**Comment by zone117x** [comment]
@hstove 
I'm in agreement about the naming for contract function invocation related RPC methods. 
There are three related actions where it would be nice to have clear and consistent names for:
* Deploying a contract [a transaction]
* Mutable contract function invocation [a transaction]
* Immutable contract function invocation [not a transaction]

An ugly but consistent set of method names would be something like: `contract_deploy_tx`, `contract_invoke_tx`, `contract_invoke_readonly`. 

Note: the current local dev CLI does not currently have an explicit command for "readonly contract function call". Rather, it has a generic `eval` command that takes `contract name` and a "raw" Clarity expression. We should implement the more limited "call" command which takes (`contract name`, `function name`, `args...`). 

> Do we have the concept of passing a value, similar to ETH?

We don't have the equivalent of this implemented yet, at least in the local dev CLI. 

> How can you get a parser error here?

I think this depends on how we perform encoding on the function args. I think a parser error here would occur if a given arg cannot be parsed as the function input type. But it probably would not be the same type of parser error that the Clarity lexer outputs.

**Comment by hstove** [comment]
> If the purpose of call_func is to evaluate a function with respect to the global prod context, then the client should run the VM locally but back reads to the contract DB to the node.

That sounds awfully complicated compared to the very light operation required for a node to just call the read-only function. Nodes can always authenticate their API if they want to prevent abuse.

**Comment by zone117x** [comment]
RE: `call_func` like method:

From some brainstorming with @jcnelson, at the minimum we could limit which functions are "freely" RPC-callable with constraints like:
* Only read-only functions can be called (our current savepoint & rollback implementation is resource intensive).
* Only functions under some low execution-cost threshold can be called. This is a cool feature that our statically analyzable system is capable of. 

These limitations help keep the cost of running a Blockstack node reasonable, and hopefully allowing us to avoid a centralization problem like Metamask+Infura.

**Comment by hstove** [comment]
> Only functions under some low execution-cost threshold can be called. This is a cool feature that our statically analyzable system is capable of.

Could this cost threshold be a configurable parameter?

## Other feature requests: 

`get_tx_receipt(txid)`. Not exactly sure what the result should look like.

`get_contract_txs(contract address)` - probably could come up with a better name. Ethereum dapp developers are always trying to figure out "how can I get a list of transactions on this contract", because there is no good RPC API for it. They end up using event logs which is kind of a workaround and takes up more space. It would also be great if you could use this RPC method to get transactions to/from any `principal`

**Comment by diwakergupta** [comment]
@jcnelson : do you have the bandwidth to work on this? Otherwise I will find another owner.

@blockstack/blockchain-team @blockstack/dx-team if folks have further input on this, please chime in. I'd like to move this forward 🙏

**Comment by jcnelson** [comment]
Not all of these methods can currently be implemented without significant additional engineering work.

Here's what we can do _right now_, in parallel to what I'm working on:

* `get_map_entry` (routed as `GET /v2/contracts/{:contract-address}/maps/{:clarity-key-value-hex}`)
* `get_contract_source` (routed as `GET /v2/contracts/{:contract-address}/source`)
* `dev_mine_block` (routed as `POST /v2/dev/testnet/local/mine_block`).  We'll need to iterate on this one for each phase of the testnet, but it shouldn't be too hard to patch this into the current testnet.

Here's what we will be able to do once the transaction-relay logic and mempool are ready:

* `send_transaction` (routed as `POST /v2/transactions`, body is the hex- or byte-encoded signed, serialized transaction).  This is already stubbed, but there's no handler for it.  We _could_ patch this through to the testnet mempool with minimal effort.

Here's what we will be able to do once #1313 merges:

* `get_deployment_cost` (routed as `POST /v2/contracts/cost`, body is Clarity source code, returns the normalized compute budget cost to analyze).  This one will be tricky to implement, since it's a potential DDoS vector (i.e. you can spam it with expensive-to-analyze transactions).

Here's what I don't think can be done anytime soon (or possibly at all -- some of these are arguably the job of the wallet, not the node):

* `call_func`.  This one's also pretty dangerous, since you can spam it with expensive-to-call functions.
* `deploy_contract`:  I don't think this belongs in the node -- I think this is the wallet's job.

**Comment by jcnelson** [comment]
What would be really helpful is if the @blockstack/dx-team could publish a JSON-RPC API wish-list here, so we can figure out what kinds of plumbing to prioritize.

**Comment by yknl** [comment]
After discussing with the team here's the wish list we came up with:


### Public Endpoints

`send_raw_transaction` (signed transaction) Broadcasts generic signed transaction

`get_transaction_status` (txid) → ok(accepted/rejected, # of confirmations), err(transaction_not_found)

`get_transaction_info` (txid) → ok(full deserialized transaction: type, sender, recipient etc.), err(transaction_not_found)

`get_token_transfer_cost` (unsigned token transaction) → ok(estimated transaction cost)

`get_map_entry` (contact identifier, map name, map key value) -> ok(value data), err()

`get_contract_source` (contract identifier) -> ok(source string), err(contract_not_found)

`get_contract_abi` (contract identifier) -> ok(contract ABI), err(contract_not_found) retrieves the contract ABI

### Authenticated Endpoints

`get_deployment_cost`  (unsigned token transaction) → ok(estimated deployment cost)

`call_func` (contract identifier, function name, function args) -> ok(function return data), err(function_not_found, parser, type checker, or runtime error)

**Comment by jcnelson** [comment]
Thanks @yknl!

I'm not sure what a "contract ABI" is.  Can you explain what the team expects from this endpoint?

**Comment by jcnelson** [comment]
Also, `get_transaction_info` and `get_transaction_status` are things that I'd expect the side-car to manage, because (like Bitcoin) the Stacks node software doesn't index individual transactions.  While this can be added (like Bitcoin's `txindex=1` feature), it will probably come after these other methods.  The node will soon be able to tell you whether or not a transaction is in the mempool, but it won't be able to tell you where it is within the blockchain (without a `txindex=1`-like feature).

**Comment by yknl** [comment]
> I'm not sure what a "contract ABI" is, since Clarity contracts aren't compiled (i.e. there is no ABI). Can you explain what the team expects from this endpoint?

It should be contract function ABI, I fixed it above. The idea is to retrieve the the interfaces for functions in the contract. It will be used to validate contract function call arguments when constructing the transaction. This can potentially be constructed client side given the contract source code, however there aren't any clarity parsers available in JS yet. cc @reedrosenbluth  

> Also, `get_transaction_info` and `get_transaction_status` are things that I'd expect the side-car to manage, because (like Bitcoin) the Stacks node software doesn't index individual transactions. While this can be added (like Bitcoin's `txindex=1` feature), it will probably come after these other methods. The node will soon be able to tell you whether or not a transaction is in the mempool, but it won't be able to tell you where it is within the blockchain (without a `txindex=1`-like feature).

Ok I'll remove the 2 endpoints from the list for now. Do we need to open issues for this in the sidecar? @zone117x

**Comment by jcnelson** [comment]
Cool.  This is also imminently doable, then.

**Comment by diwakergupta** [comment]
@kantai are you planning to address _all_ the endpoints as part of your changes in this sprint? If not, I suggest 1) re-wording this issue as just describing the required endpoints and then closing it; 2) opening a new issue listing out exactly the endpoints you're currently planning to implement.

**Comment by kantai** [comment]
> @kantai are you planning to address all the endpoints as part of your changes in this sprint? If not, I suggest 1) re-wording this issue as just describing the required endpoints and then closing it; 2) opening a new issue listing out exactly the endpoints you're currently planning to implement.

Yep -- the PR will implement the following:

* send_raw_transaction (signed transaction) Broadcasts generic signed transaction -- this will not actually accept the data, and instead reject all requests. full implementation will require the mempool implementation.

* get_token_transfer_cost (unsigned token transaction) → ok(estimated transaction cost)
* get_deployment_cost (unsigned token transaction) → ok(estimated deployment cost)

both of these will be returning very simple costs for now -- basically just the minimum cost

* get_map_entry (contact identifier, map name, map key value) -> ok(value data), err()
* get_contract_source (contract identifier) -> ok(source string), err(contract_not_found)
*  get_contract_abi (contract identifier, function name) -> ok(contract ABI), err(contract_not_found)
* call_func (contract identifier, function name, function args) -> ok(function return data), err(function_not_found, parser, type checker, or runtime error)

**Comment by friedger** [comment]
What about `get_var` and `get_constant` ?

**Comment by yknl** [comment]
> What about get_var and get_constant ?

Can you explain the use case?

**Comment by jcnelson** [comment]
Getting a var makes sense to me.  The code path is very similar to the path for querying a data map key, and this is state that we track in the MARF anyway.

Getting a constant makes less sense, since (1) they're not stored anywhere, and (2) they can be extracted from the contract ABI already.

**Comment by kantai** [comment]
Seeing as this issue is going to be tied to #1382 -- maybe we should open new issues for any additional endpoints we want to implement? That would be:

* get_deployment_cost: #1398
* get_var: #1399
* get_constant: #1400

And then we can discuss the endpoints there

**Comment by njordhov** [comment]
> The whole state of the chain is public knowledge, so it doesn't make much sense to constrain the RESTful APIs from allowing read-only access to them.

I would like to structure clarity contracts with *all mutating calls* lifted to public functions. See #1393 and #1334. Accordingly, private user functions won't mutate data and thus could be exposed as read-only, no?

**Comment by jcnelson** [comment]
> I would like to structure clarity contracts with all mutating calls lifted to public user functions.

You are of course free to do this in your own contracts, but this does not seem like something Clarity should enforce.

**Comment by kantai** [comment]
Per my comment above -- I'm closing this issue with PR #1382 -- for discussion of other endpoints we may want, we should use those specific issues.

**Comment by njordhov** [comment]
> this does not seem like something Clarity should enforce
@jcnelson I am only pointing out a potential benefit of lifting mutations to the public functions: All other functions will be read only and can thus be exposed rather than be kept private, as they don't mutate data. Clarity can leverage this without enforcing immutability in private functions.
